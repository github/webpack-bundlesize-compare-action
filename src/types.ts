export type AssetDiff = {
  name: string
  new: {
    size: number
    gzipSize: number | null
  }
  old: {
    size: number
    gzipSize: number | null
  }
  diff: number
  diffPercentage: number
}

export interface Sizes {
  size: number
  gzipSize: number | null
}

export interface ChunkSizes extends Sizes {
  initial: boolean
}

export function isChunkSizes(sizes: Sizes): sizes is ChunkSizes {
  return 'initial' in sizes
}

interface _WebpackStatsDiff {
  added: AssetDiff[]
  removed: AssetDiff[]
  bigger: AssetDiff[]
  smaller: AssetDiff[]
  unchanged: AssetDiff[]
  total: AssetDiff
  initial?: AssetDiff
}

export type WebpackStatsDiff<T extends Sizes> = T extends ChunkSizes
  ? _WebpackStatsDiff
  : Omit<_WebpackStatsDiff, 'initial'>
