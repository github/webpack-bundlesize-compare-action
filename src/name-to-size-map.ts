import {StatsCompilation} from 'webpack'
import type {Sizes, ChunkSizes} from './types'

export function assetNameToSizeMap(
  statAssets: StatsCompilation['assets'] = []
): Map<string, Sizes> {
  return new Map(
    statAssets
      // when Webpack's stats.excludeAssets is used, assets which are excluded will be grouped into an asset with type 'hidden assets'
      .filter(it => !!it.name && it.type !== 'hidden assets')
      .map(asset => {
        let gzipSize: number | null = null
        if (asset.related && Array.isArray(asset.related)) {
          const gzipAsset = asset.related.find(
            related => related.type === 'gzipped'
          )
          if (gzipAsset) {
            gzipSize = gzipAsset.size
          }
        }

        return [
          asset.name,
          {
            size: asset.size,
            gzipSize
          }
        ]
      })
  )
}

export function chunkModuleNameToSizeMap(
  statChunks: StatsCompilation['chunks'] = []
): Map<string, ChunkSizes> {
  return new Map(
    statChunks.flatMap(chunk => {
      if (!chunk.modules) return []
      return chunk.modules.flatMap(module => {
        // If a module doesn't have any submodules beneath it, then just return its own size
        // Otherwise, break each module into its submodules with their own sizes
        if (module.modules) {
          return module.modules.map(submodule => [
            submodule.name ?? '',
            {
              initial: chunk.initial,
              size: submodule.size ?? 0,
              gzipSize: null
            }
          ])
        } else {
          return [
            [
              module.name ?? '',
              {
                initial: chunk.initial,
                size: module.size ?? 0,
                gzipSize: null
              }
            ]
          ]
        }
      })
    })
  )
}
