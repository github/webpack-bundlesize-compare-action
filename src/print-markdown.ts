import {formatFileSizeIEC} from './file-sizes'
import type {AssetDiff, WebpackStatsDiff} from './types'

function conditionalPercentage(number: number): string {
  if ([Infinity, -Infinity].includes(number)) {
    return '-'
  }

  const absValue = Math.abs(number)

  if ([0, 100].includes(absValue)) {
    return `${number}%`
  }

  const value = [0, 100].includes(absValue) ? absValue : absValue.toFixed(2)

  return `${signFor(number)}${value}%`
}

function capitalize(text: string): string {
  return `${text[0].toUpperCase()}${text.slice(1)}`
}

function makeHeader(columns: readonly string[]): string {
  return `${columns.join(' | ')}
${columns
  .map(x =>
    Array.from(new Array(x.length))
      .map(() => '-')
      .join('')
  )
  .join(' | ')}`
}

const TOTAL_HEADERS = makeHeader([
  'Files count',
  'Total bundle size',
  '% Changed'
])
const TABLE_HEADERS = makeHeader(['Asset', 'File Size', '% Changed'])

function signFor(num: number): '' | '+' | '-' {
  if (num === 0) return ''
  return num > 0 ? '+' : '-'
}

function toFileSizeDiff(
  oldSize: number | null,
  newSize: number | null,
  diff?: number | undefined
): string {
  const diffLine = [
    `${formatFileSizeIEC(oldSize)} â†’ ${formatFileSizeIEC(newSize)}`
  ]
  if (typeof diff !== 'undefined') {
    diffLine.push(`(${signFor(diff)}${formatFileSizeIEC(diff)})`)
  }
  return diffLine.join(' ')
}

function toFileSizeDiffCell(asset: AssetDiff): string {
  const lines = []
  if (asset.diff === 0) {
    lines.push(formatFileSizeIEC(asset.new.size))
    if (asset.new.gzipSize) {
      lines.push(formatFileSizeIEC(asset.new.gzipSize))
    }
  } else {
    lines.push(toFileSizeDiff(asset.old.size, asset.new.size, asset.diff))
    if (asset.old.gzipSize || asset.new.gzipSize) {
      lines.push(
        `${toFileSizeDiff(asset.old.gzipSize, asset.new.gzipSize)} (gzip)`
      )
    }
  }

  return lines.join('<br />')
}

function printAssetTableRow(asset: AssetDiff): string {
  return [
    asset.name,
    toFileSizeDiffCell(asset),
    conditionalPercentage(asset.diffPercentage)
  ].join(' | ')
}

export function printAssetTablesByGroup(
  statsDiff: Omit<WebpackStatsDiff, 'total'>
): string {
  const statsFields = [
    'added',
    'removed',
    'bigger',
    'smaller',
    'unchanged'
  ] as const
  return statsFields
    .map(field => {
      const assets = statsDiff[field]
      if (assets.length === 0) {
        return `**${capitalize(field)}**

No assets were ${field}`
      }

      return `**${capitalize(field)}**

${TABLE_HEADERS}
${assets
  .map(asset => {
    return printAssetTableRow(asset)
  })
  .join('\n')}`
    })
    .join('\n\n')
}

const getDiffEmoji = (diff: AssetDiff): string =>
  diff.diffPercentage === Infinity
    ? 'ðŸ†•'
    : diff.diffPercentage <= -100
      ? 'ðŸ”¥'
      : diff.diffPercentage > 0
        ? 'ðŸ“ˆ'
        : diff.diffPercentage < 0
          ? 'ðŸ“‰'
          : ' '

const getTrimmedChunkName = (chunkModule: AssetDiff): string => {
  let chunkName = chunkModule.name
  if (chunkName.startsWith('./')) {
    chunkName = chunkName.substring(2)
  } else if (chunkName.startsWith('/')) {
    chunkName = chunkName.substring(1)
  }
  return chunkName
}

const CHUNK_TABLE_HEADERS = makeHeader(['File', 'Î”', 'Size'])

function printChunkModuleRow(chunkModule: AssetDiff): string {
  const emoji = getDiffEmoji(chunkModule)
  const chunkName = getTrimmedChunkName(chunkModule)

  return [
    `\`${chunkName}\``,
    `${emoji} ${chunkModule.diff >= 0 ? '+' : '-'}${formatFileSizeIEC(
      chunkModule.diff
    )}${
      Number.isFinite(chunkModule.diffPercentage)
        ? ` (${conditionalPercentage(chunkModule.diffPercentage)})`
        : ''
    }`,
    `${formatFileSizeIEC(chunkModule.old.size)} â†’ ${formatFileSizeIEC(
      chunkModule.new.size
    )}`
  ].join(' | ')
}

export function printChunkModulesTable(
  statsDiff: Omit<WebpackStatsDiff, 'total' | 'unchanged'> | null
): string {
  if (!statsDiff) return ''
  const changedModules = [
    ...statsDiff.added,
    ...statsDiff.removed,
    ...statsDiff.bigger,
    ...statsDiff.smaller
  ].sort((a, b) => b.diffPercentage - a.diffPercentage)

  if (changedModules.length === 0) {
    return `
Changeset

No files were changed`
  }

  return `
<details>
<summary>Changeset${
    changedModules.length > 100 ? ' (largest 100 files by percent change)' : ''
  }</summary>

${CHUNK_TABLE_HEADERS}
${changedModules
  .slice(0, 100)
  .map(chunkModule => printChunkModuleRow(chunkModule))
  .join('\n')}

</details>
`
}

export function printTotalAssetTable(
  statsDiff: Pick<WebpackStatsDiff, 'total'>
): string {
  return `**Total**

${TOTAL_HEADERS}
${printAssetTableRow(statsDiff.total)}`
}
