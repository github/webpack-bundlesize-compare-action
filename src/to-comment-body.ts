import {
  printAssetTablesByGroup,
  printChunkModulesTable,
  printTotalAssetTable
} from './print-markdown'
import type {WebpackStatsDiff, Sizes, ChunkSizes} from './types'

export function getIdentifierComment(key: string): string {
  return `<!--- bundlestats-action-comment${key ? ` key:${key}` : ''} --->`
}

export function getCommentBody(
  statsDiff: WebpackStatsDiff<Sizes>,
  chunkModuleDiff: WebpackStatsDiff<ChunkSizes> | undefined,
  title: string
): string {
  return `
### Bundle Stats${title ? ` - ${title}` : ''}

Hey there, this message comes from a [GitHub action](https://github.com/github/webpack-bundlesize-compare-action) that helps you and reviewers to understand how these changes affect the size of this project's bundle.

As this PR is updated, I'll keep you updated on how the bundle size is impacted.

${printTotalAssetTable(statsDiff, chunkModuleDiff)}
${chunkModuleDiff ? `${printChunkModulesTable(chunkModuleDiff)}\n` : ''}
<details>
<summary>View detailed bundle breakdown</summary>

<div>

${printAssetTablesByGroup(statsDiff)}

</div>
</details>

${getIdentifierComment(title)}
`
}
