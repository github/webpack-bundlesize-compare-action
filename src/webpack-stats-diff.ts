import {getAssetDiff} from './get-asset-diff'
import {sortDiffDescending} from './sort-diff-descending'
import {Sizes, WebpackStatsDiff, isChunkSizes} from './types'

export function webpackStatsDiff<T extends Sizes>(
  oldAssets: Map<string, T>,
  newAssets: Map<string, T>
): WebpackStatsDiff<T> {
  const added = []
  const removed = []
  const bigger = []
  const smaller = []
  const unchanged = []

  let newSizeTotal = 0
  let oldSizeTotal = 0
  let newGzipSizeTotal = 0
  let oldGzipSizeTotal = 0

  let newSizeInitial = 0
  let oldSizeInitial = 0
  let newGzipSizeInitial = 0
  let oldGzipSizeInitial = 0
  let newInitialFilesCount = 0
  let oldInitialFilesCount = 0
  let hasInitialData = false

  for (const [name, oldAssetSizes] of oldAssets) {
    oldSizeTotal += oldAssetSizes.size
    oldGzipSizeTotal += oldAssetSizes.gzipSize ?? NaN
    if (isChunkSizes(oldAssetSizes) && oldAssetSizes.initial) {
      hasInitialData = true
      oldSizeInitial += oldAssetSizes.size
      oldGzipSizeInitial += oldAssetSizes.gzipSize ?? NaN
      oldInitialFilesCount += 1
    }
    const newAsset = newAssets.get(name)
    if (!newAsset) {
      removed.push(getAssetDiff(name, oldAssetSizes, {size: 0, gzipSize: 0}))
    } else {
      const diff = getAssetDiff(name, oldAssetSizes, newAsset)

      if (diff.diffPercentage > 0) {
        bigger.push(diff)
      } else if (diff.diffPercentage < 0) {
        smaller.push(diff)
      } else {
        unchanged.push(diff)
      }
    }
  }

  for (const [name, newAssetSizes] of newAssets) {
    newSizeTotal += newAssetSizes.size
    newGzipSizeTotal += newAssetSizes.gzipSize ?? NaN
    if (isChunkSizes(newAssetSizes) && newAssetSizes.initial) {
      hasInitialData = true
      newSizeInitial += newAssetSizes.size
      newGzipSizeInitial += newAssetSizes.gzipSize ?? NaN
      newInitialFilesCount += 1
    }
    const oldAsset = oldAssets.get(name)
    if (!oldAsset) {
      added.push(getAssetDiff(name, {size: 0, gzipSize: 0}, newAssetSizes))
    }
  }

  const oldFilesCount = oldAssets.size
  const newFilesCount = newAssets.size
  return {
    added: sortDiffDescending(added),
    removed: sortDiffDescending(removed),
    bigger: sortDiffDescending(bigger),
    smaller: sortDiffDescending(smaller),
    unchanged,
    total: getAssetDiff(
      oldFilesCount === newFilesCount
        ? `${newFilesCount}`
        : `${oldFilesCount} -> ${newFilesCount}`,
      {size: oldSizeTotal, gzipSize: oldGzipSizeTotal},
      {size: newSizeTotal, gzipSize: newGzipSizeTotal}
    ),
    initial: hasInitialData
      ? getAssetDiff(
          oldInitialFilesCount === newInitialFilesCount
            ? `(initial only) ${newInitialFilesCount}`
            : `(initial only) ${oldInitialFilesCount} -> ${newInitialFilesCount}`,
          {size: oldSizeInitial, gzipSize: oldGzipSizeInitial},
          {size: newSizeInitial, gzipSize: newGzipSizeInitial}
        )
      : undefined
  } as WebpackStatsDiff<T>
}
