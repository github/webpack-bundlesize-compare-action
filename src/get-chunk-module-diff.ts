import type {StatsCompilation} from 'webpack'
import {chunkModuleNameToSizeMap} from './name-to-size-map'
import type {WebpackStatsDiff, ChunkSizes} from './types'
import {webpackStatsDiff} from './webpack-stats-diff'

export function getChunkModuleDiff(
  oldStats: Pick<StatsCompilation, 'chunks'>,
  newStats: Pick<StatsCompilation, 'chunks'>
): WebpackStatsDiff<ChunkSizes> | undefined {
  if (!oldStats.chunks || !newStats.chunks) {
    return undefined
  }
  return webpackStatsDiff(
    chunkModuleNameToSizeMap(oldStats.chunks),
    chunkModuleNameToSizeMap(newStats.chunks)
  )
}
