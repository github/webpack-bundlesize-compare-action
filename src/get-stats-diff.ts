import type {StatsCompilation} from 'webpack'
import {assetNameToSizeMap} from './name-to-size-map'
import type {WebpackStatsDiff, Sizes} from './types'
import {webpackStatsDiff} from './webpack-stats-diff'

export function getStatsDiff(
  oldAssetStats: Pick<StatsCompilation, 'assets'>,
  newAssetStats: Pick<StatsCompilation, 'assets'>
): WebpackStatsDiff<Sizes> {
  return webpackStatsDiff(
    assetNameToSizeMap(oldAssetStats.assets),
    assetNameToSizeMap(newAssetStats.assets)
  )
}
