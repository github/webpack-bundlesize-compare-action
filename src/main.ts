import * as core from '@actions/core'
import {context, getOctokit} from '@actions/github'
import {getChunkModuleDiff} from './get-chunk-module-diff'
import {getStatsDiff} from './get-stats-diff'
import {parseStatsFileToJson} from './parse-stats-file-to-json'
import {getCommentBody, getIdentifierComment} from './to-comment-body'
import {
  isDescribeAssetsSection,
  DescribeAssetsSection,
  DescribeAssetsOptions
} from './types'

export function getDescribeAssetsOptions(
  optionString: string
): DescribeAssetsOptions {
  optionString = optionString.trim().toLowerCase()
  if (optionString === 'all') {
    optionString = 'bigger smaller added removed unchanged'
  } else if (optionString === 'none') {
    optionString = ''
  } else if (optionString === 'changed-only') {
    optionString = 'bigger smaller added removed'
  }
  const options = {
    added: false,
    removed: false,
    bigger: false,
    smaller: false,
    unchanged: false
  }
  if (optionString) {
    const sections = optionString.split(' ').filter(s => !!s)
    if (sections.some(s => !isDescribeAssetsSection(s))) {
      throw new Error(
        `Unsupported options for 'describe-assets': '${optionString}' contains unexpected sections`
      )
    }
    for (const s of sections as DescribeAssetsSection[]) {
      options[s] = true
    }
  }
  return options
}

async function run(): Promise<void> {
  try {
    if (
      context.eventName !== 'pull_request' &&
      context.eventName !== 'pull_request_target'
    ) {
      throw new Error(
        'This action only supports pull_request and pull_request_target events'
      )
    }
    const {
      issue: {number: issue_number},
      repo: {owner, repo: repo_name}
    } = context
    const token = core.getInput('github-token')
    const currentStatsJsonPath = core.getInput('current-stats-json-path')
    const baseStatsJsonPath = core.getInput('base-stats-json-path')
    const describeAssetsOptionString = core.getInput('describe-assets')
    const describeAssetsOptions = getDescribeAssetsOptions(
      describeAssetsOptionString
    )
    const title = core.getInput('title') ?? ''
    const {rest} = getOctokit(token)

    const [currentStatsJson, baseStatsJson, {data: comments}] =
      await Promise.all([
        parseStatsFileToJson(currentStatsJsonPath),
        parseStatsFileToJson(baseStatsJsonPath),
        rest.issues.listComments({
          repo: repo_name,
          owner,
          issue_number
        })
      ])

    const identifierComment = getIdentifierComment(title)

    const [currentComment, ...restComments] = comments.filter(
      comment =>
        comment.user?.login === 'github-actions[bot]' &&
        comment.body &&
        comment.body.includes(identifierComment)
    )

    const statsDiff = getStatsDiff(baseStatsJson, currentStatsJson)
    const chunkModuleDiff = getChunkModuleDiff(baseStatsJson, currentStatsJson)

    const commentBody = getCommentBody(
      statsDiff,
      chunkModuleDiff,
      title,
      describeAssetsOptions
    )

    const promises: Array<Promise<unknown>> = []

    if (restComments.length > 1) {
      promises.push(
        ...restComments.map(async comment => {
          return rest.issues.deleteComment({
            repo: repo_name,
            owner,
            comment_id: comment.id
          })
        })
      )
    }

    if (currentComment) {
      promises.push(
        rest.issues.updateComment({
          issue_number,
          owner,
          repo: repo_name,
          body: commentBody,
          comment_id: currentComment.id
        })
      )
    } else {
      promises.push(
        rest.issues.createComment({
          issue_number,
          owner,
          repo: repo_name,
          body: commentBody
        })
      )
    }

    await Promise.all(promises)
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

// Only run if this module is being executed directly, not imported
if (require.main === module) {
  run()
}
