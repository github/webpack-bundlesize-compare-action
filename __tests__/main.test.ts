import {expect, test} from '@jest/globals'
import {getStatsDiff} from '../src/get-stats-diff'
import {getChunkModuleDiff} from '../src/get-chunk-module-diff'
import {
  printAssetTablesByGroup,
  printChunkModulesTable,
  printTotalAssetTable
} from '../src/print-markdown'
import {AssetDiff} from '../src/types'
import {readFile} from 'node:fs/promises'
import {resolve} from 'node:path'
import {StatsCompilation} from 'webpack'

async function readJsonFile(path: string): Promise<StatsCompilation> {
  const data = await readFile(resolve(__dirname, path), 'utf8')
  return JSON.parse(data)
}

test('Shows stats when files are removed', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/old-stats-assets.json'),
    await readJsonFile('./__mocks__/new-stats-assets.json')
  )

  expect(printTotalAssetTable(statsDiff)).toMatchSnapshot()
  expect(printAssetTablesByGroup(statsDiff)).toMatchSnapshot()
})

test('Shows stats when files are added', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/new-stats-assets.json'),
    await readJsonFile('./__mocks__/old-stats-assets.json')
  )

  expect(printTotalAssetTable(statsDiff)).toMatchSnapshot()
  expect(printAssetTablesByGroup(statsDiff)).toMatchSnapshot()
})

test('Shows stats when files are unchanged', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/old-stats-assets.json'),
    await readJsonFile('./__mocks__/old-stats-assets.json')
  )

  expect(printTotalAssetTable(statsDiff)).toMatchSnapshot()
  expect(printAssetTablesByGroup(statsDiff)).toMatchSnapshot()
})

test('Shows stats when files are hidden', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/old-hidden-stats-assets.json'),
    await readJsonFile('./__mocks__/new-hidden-stats-assets.json')
  )

  expect(printTotalAssetTable(statsDiff)).toMatchSnapshot()
  expect(printAssetTablesByGroup(statsDiff)).toMatchSnapshot()
})

test('computes the correct module diff information', async () => {
  const statsDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-with-chunks.json'),
    await readJsonFile('./__mocks__/new-stats-with-chunks.json')
  )

  expect(statsDiff?.added).toContainEqual({
    name: './src/client/this-file-was-added.ts',
    diff: 1496,
    diffPercentage: Infinity,
    new: {size: 1496, gzipSize: NaN},
    old: {size: 0, gzipSize: 0}
  } as AssetDiff)
  expect(statsDiff?.bigger).toContainEqual({
    name: './src/client/this-file-grew-larger.tsx',
    diff: 200,
    diffPercentage: 35.58719,
    new: {size: 762, gzipSize: NaN},
    old: {size: 562, gzipSize: NaN}
  } as AssetDiff)
  expect(statsDiff?.smaller).toContainEqual({
    name: './src/client/helpers/this-file-grew-smaller.ts',
    diff: -200,
    diffPercentage: -7.94281,
    new: {size: 2318, gzipSize: NaN},
    old: {size: 2518, gzipSize: NaN}
  } as AssetDiff)
  expect(statsDiff?.removed).toContainEqual({
    name: './src/client/this-file-will-be-deleted.ts',
    diff: -1496,
    diffPercentage: -100,
    new: {size: 0, gzipSize: 0},
    old: {size: 1496, gzipSize: NaN}
  } as AssetDiff)
  expect(statsDiff?.total.new).toEqual(statsDiff?.total.old)
  expect(statsDiff?.total.diff).toEqual(0)
  expect(statsDiff?.total.diffPercentage).toEqual(0)
  // No async chunks results in no initial stats being reported
  expect(statsDiff?.initial).toBe(undefined)

  // TODO(mariaines): add test case for initial stats for a build with async chunks
})

// TODO(mariaines): update this snapshot for testing initial stats
test('Shows initial stats for builds with async chunks', async () => {
  const statsDiff = getStatsDiff(
    await readJsonFile('./__mocks__/old-stats-assets.json'),
    await readJsonFile('./__mocks__/new-stats-assets.json')
  )
  const chunkModuleDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-with-chunks.json'),
    await readJsonFile('./__mocks__/new-stats-with-chunks.json')
  )
  expect(printTotalAssetTable(statsDiff, chunkModuleDiff)).toMatchSnapshot()
})

test('displays module information when files are added/removed/changed', async () => {
  const statsDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-with-chunks.json'),
    await readJsonFile('./__mocks__/new-stats-with-chunks.json')
  )

  expect(printChunkModulesTable(statsDiff)).toMatchSnapshot()
})

test('displays no module information when unchanged', async () => {
  const statsDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-with-chunks.json'),
    await readJsonFile('./__mocks__/old-stats-with-chunks.json')
  )

  expect(printChunkModulesTable(statsDiff)).toMatchSnapshot()
})

test('does not display module information when it does not exist', async () => {
  const statsDiff = getChunkModuleDiff(
    await readJsonFile('./__mocks__/old-stats-assets.json'),
    await readJsonFile('./__mocks__/old-stats-assets.json')
  )

  expect(printChunkModulesTable(statsDiff)).toMatchSnapshot()
})
