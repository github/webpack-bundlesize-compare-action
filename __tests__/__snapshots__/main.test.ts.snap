// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shows initial stats for builds with async chunks 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
7 -> 2 | 1.34 MB -> 1.29 MB (-53.65 KB)<br />386.44 KB -> N/A (gzip) | -3.91%"
`;

exports[`Shows stats when files are added 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
2 -> 7 | 1.29 MB -> 1.34 MB (+53.65 KB)<br />N/A -> 386.44 KB (gzip) | +4.07%"
`;

exports[`Shows stats when files are added 2`] = `
"**Added**

Asset | File Size | % Changed
----- | --------- | ---------
296.chunk.js | 0 B -> 124.57 KB (+124.57 KB)<br />0 B -> 35.05 KB (gzip) | -
288.chunk.js | 0 B -> 57.24 KB (+57.24 KB)<br />0 B -> 16.33 KB (gzip) | -
920.chunk.js | 0 B -> 54.98 KB (+54.98 KB)<br />0 B -> 17.08 KB (gzip) | -
912.chunk.js | 0 B -> 44.37 KB (+44.37 KB)<br />0 B -> 14.31 KB (gzip) | -
699.chunk.js | 0 B -> 26.39 KB (+26.39 KB)<br />0 B -> 6.14 KB (gzip) | -

**Removed**

No assets were removed

**Bigger**

Asset | File Size | % Changed
----- | --------- | ---------
manifest.json | 91 B -> 551 B (+460 B)<br />N/A -> 151 B (gzip) | +505.49%

**Smaller**

Asset | File Size | % Changed
----- | --------- | ---------
app.bundle.js | 1.29 MB -> 1.04 MB (-254.35 KB)<br />N/A -> 297.38 KB (gzip) | -19.29%

**Unchanged**

No assets were unchanged"
`;

exports[`Shows stats when files are hidden 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
0 | 0 B | 0%"
`;

exports[`Shows stats when files are hidden 2`] = `
"**Added**

No assets were added

**Removed**

No assets were removed

**Bigger**

No assets were bigger

**Smaller**

No assets were smaller

**Unchanged**

No assets were unchanged"
`;

exports[`Shows stats when files are removed 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
7 -> 2 | 1.34 MB -> 1.29 MB (-53.65 KB)<br />386.44 KB -> N/A (gzip) | -3.91%"
`;

exports[`Shows stats when files are removed 2`] = `
"**Added**

No assets were added

**Removed**

Asset | File Size | % Changed
----- | --------- | ---------
296.chunk.js | 124.57 KB -> 0 B (-124.57 KB)<br />35.05 KB -> 0 B (gzip) | -100%
288.chunk.js | 57.24 KB -> 0 B (-57.24 KB)<br />16.33 KB -> 0 B (gzip) | -100%
920.chunk.js | 54.98 KB -> 0 B (-54.98 KB)<br />17.08 KB -> 0 B (gzip) | -100%
912.chunk.js | 44.37 KB -> 0 B (-44.37 KB)<br />14.31 KB -> 0 B (gzip) | -100%
699.chunk.js | 26.39 KB -> 0 B (-26.39 KB)<br />6.14 KB -> 0 B (gzip) | -100%

**Bigger**

Asset | File Size | % Changed
----- | --------- | ---------
app.bundle.js | 1.04 MB -> 1.29 MB (+254.35 KB)<br />297.38 KB -> N/A (gzip) | +23.91%

**Smaller**

Asset | File Size | % Changed
----- | --------- | ---------
manifest.json | 551 B -> 91 B (-460 B)<br />151 B -> N/A (gzip) | -83.48%

**Unchanged**

No assets were unchanged"
`;

exports[`Shows stats when files are unchanged 1`] = `
"**Total**

Files count | Total bundle size | % Changed
----------- | ----------------- | ---------
7 | 1.34 MB<br />386.44 KB | 0%"
`;

exports[`Shows stats when files are unchanged 2`] = `
"**Added**

No assets were added

**Removed**

No assets were removed

**Bigger**

No assets were bigger

**Smaller**

No assets were smaller

**Unchanged**

Asset | File Size | % Changed
----- | --------- | ---------
app.bundle.js | 1.04 MB<br />297.38 KB | 0%
296.chunk.js | 124.57 KB<br />35.05 KB | 0%
288.chunk.js | 57.24 KB<br />16.33 KB | 0%
920.chunk.js | 54.98 KB<br />17.08 KB | 0%
912.chunk.js | 44.37 KB<br />14.31 KB | 0%
699.chunk.js | 26.39 KB<br />6.14 KB | 0%
manifest.json | 551 B<br />151 B | 0%"
`;

exports[`displays module information when files are added/removed/changed 1`] = `
"
<details>
<summary>Changeset</summary>

File | Î” | Size
---- | - | ----
\`src/client/this/file/has/a/long/name/so-that-it-will-hopefully-wrap.tsx\` | ðŸ†• +2.72 KB | 0 B -> 2.72 KB
\`src/client/this-file-was-added.ts\` | ðŸ†• +1.46 KB | 0 B -> 1.46 KB
\`src/client/this-is-another-file-with-a-long-name-to-test-the-table-sizing.tsx\` | ðŸ†• +1.04 KB | 0 B -> 1.04 KB
\`src/client/helpers/this-file-is-much-larger.ts\` | ðŸ“ˆ +11.66 KB (+865.77%) | 1.35 KB -> 13.01 KB
\`src/client/this-file-grew-larger.tsx\` | ðŸ“ˆ +200 B (+35.59%) | 562 B -> 762 B
\`src/client/helpers/this-file-grew-smaller.ts\` | ðŸ“‰ -200 B (-7.94%) | 2.46 KB -> 2.26 KB
\`src/client/this-file-is-much-smaller.tsx\` | ðŸ“‰ -11.66 KB (-99.92%) | 11.67 KB -> 10 B
\`src/client/routes.tsx\` | ðŸ”¥ -2.72 KB (-100%) | 2.72 KB -> 0 B
\`src/client/this-file-will-be-deleted.ts\` | ðŸ”¥ -1.46 KB (-100%) | 1.46 KB -> 0 B
\`src/client/render-memex.tsx\` | ðŸ”¥ -1.04 KB (-100%) | 1.04 KB -> 0 B

</details>
"
`;

exports[`displays no module information when unchanged 1`] = `
"
Changeset

No files were changed"
`;

exports[`does not display module information when it does not exist 1`] = `""`;
